src/mongoc/mongoc-stream-tls.c

```diff
@@ -495,40 +497,131 @@ _mongoc_stream_tls_writev (mongoc_stream_t *stream,
       iov_pos = 0;
 
       while (iov_pos < iov[i].iov_len) {
-         write_ret = BIO_write (tls->bio, (char *)iov[i].iov_base + iov_pos,
-                              (int)(iov[i].iov_len - iov_pos));
+         if (buf && (buf_head != buf_tail ||
+             ((i + 1 < iovcnt) && ((buf_end - buf_tail) > (iov[i].iov_len - iov_pos))))) {
```

Comment:
If you haven't satisfied the first part of the or clause (```buf_head != buf_tail```) then you have not buffered anything yet.  If the size of the iov is equal to the buffer length, then you'd just end up buffering the message and sending it immediately because it filled the buffer, so don't bother buffering it.